### PR TITLE
Open and Fix NativeTest by Spring Framework under GraalVM CE

### LINKED ISSUE #465: 添加对spring native的支持
image

# Enviroment

JDK Version(required): graalvm-ce-java11-22.3.0

SpringBoot Version(required): 2.7.5

dynamic-datasource-spring-boot-starter Version(required):3.5.2

druid Version(optional):

# Describe what happened
Native Build Tools构建本地可执行文件时出现

Expected Result:

Actual Result:

If there is an exception,or aop invalid,please attach the exception trace:

```
Just paste your stack trace here!
```

# Steps to reproduce

- Step 1 

- Step 2 

- Step 3

### LINKED ISSUE #535: Graalvm can compile and package normally, but cannot load the data source after startup
Please fill it out carefully, or it will be closed.
请认真填写，不然会直接关闭。

# Enviroment
Window10
JDK Version(required): 
java version "17.0.7" 2023-04-18 LTS
SpringBoot Version(required): 
3.1.1
dynamic-datasource-spring-boot-starter Version(required):
4.1.1
druid Version(optional):
java version "17.0.7" 2023-04-18 LTS

# Describe what happened
graalvm可正常编译打包，启动后加载不到数据源

Expected Result:
image

Actual Result:

If there is an exception,or aop invalid,please attach the exception trace:

```
Just paste your stack trace here!
```

# Steps to reproduce

- Step 1 

- Step 2 

- Step 3

### PR DESCRIPTION
**Pull request by linghengqian**
**What kind of change does this PR introduce?** (check at least one)

- [x] Bugfix
- [ ] Feature
- [ ] Code style
- [ ] Refactor
- [ ] Doc
- [ ] Other, please describe:

**The description of the PR:**

- Fixes #465.
- Fixes #535.
- Open and Fix NativeTest by Spring Framework under GraalVM CE.

**Other information:**

- `DynamicRoutingDataSource` has design flaws. This class attempts to inject a `List<DynamicDataSourceProvider>` Spring Bean into a non-container environment, which causes the Spring Framework to deliberately take over this class and set a NoArgConstructor for this class with an unaligned parameter list. 

- As a fix, I added a AllArgsConstructor for `DynamicRoutingDataSource` to avoid https://github.com/baomidou/dynamic-datasource/actions/runs/6080472779/job/16494489274?pr=560 display, injected `List<DynamicDataSourceProvider>`  is null. 

- Similarly, Alibaba Druid's Spring Boot Starter has a similar problem with Spring Bean introspection failures, resulting in the Spring Framework's improper initialization of beans.

### CODE REVIEW

File: dynamic-datasource-spring/src/main/java/com/baomidou/dynamic/datasource/DynamicRoutingDataSource.java [hunk 1]
```diff
@@ -48,6 +49,7 @@
  * @since 1.0.0
  */
 @Slf4j
+@Component
 public class DynamicRoutingDataSource extends AbstractRoutingDataSource implements InitializingBean, DisposableBean {
```
[the remark this answers is not in the document]
Comment (linghengqian) [reply dynamic-datasource-spring/src/main/java/com/baomidou/dynamic/datasource/DynamicRoutingDataSource.java:53]: - @huayanYu  I haven't finished this PR yet, I'm just debugging the configuration of Github Actions.
- It's not this place that really needs attention. When I confirm the direction change, the change should be removed.

[the remark this answers is not in the document]
Comment (huayanYu) [reply dynamic-datasource-spring/src/main/java/com/baomidou/dynamic/datasource/DynamicRoutingDataSource.java:53]: ok i already up spring boot version to 2.7.15, and test in ds-spring-boot went wrong.

[the remark this answers is not in the document]
Comment (linghengqian) [reply dynamic-datasource-spring/src/main/java/com/baomidou/dynamic/datasource/DynamicRoutingDataSource.java:53]: > ok i already up spring boot version to 2.7.15, and test in ds-spring-boot went wrong. 

- According to the CI log, Spring Boot OSS 2 does not include Junit4 classes.  This requires some refactoring procedures in another module test.

- I will fix it later. I temporarily leave my desk.

[the remark this answers is not in the document]
Comment (linghengqian) [reply dynamic-datasource-spring/src/main/java/com/baomidou/dynamic/datasource/DynamicRoutingDataSource.java:53]: - Refer to https://github.com/baomidou/dynamic-datasource/pull/564 .

**Comment by linghengqian** [comment]
- Call @huayanYu to help review.

- `DynamicRoutingDataSource` has design flaws. This class attempts to inject a `List<DynamicDataSourceProvider>` Spring Bean into a non-container environment, which causes the Spring Framework to deliberately take over this class and set a NoArgConstructor for this class with an unaligned parameter list. 

- As a fix, I added a AllArgsConstructor for `DynamicRoutingDataSource` to avoid https://github.com/baomidou/dynamic-datasource/actions/runs/6080472779/job/16494489274?pr=560 display, injected `List<DynamicDataSourceProvider>`  is null. 

- Similarly, Alibaba Druid's Spring Boot Starter has a similar problem with Spring Bean introspection failures, resulting in the Spring Framework's improper initialization of beans.  

- I'll discuss how to manage GraalVM Reachability Metadata and convert Standard Metadata to Conditional Metadata in a new issue. Now the downstream still needs to add some of the contents of the JSON file itself. But the core problem has been solved, because nativeTest has already run through.

- From this point of view, apart from the fact that SPEL cannot be used, we can claim that `dynamic-datasource` should not conflict with GraalVM Native Image.

**Comment by linghengqian** [comment]
- More early discussion of the contents of this PR is at https://github.com/spring-projects/spring-framework/issues/29614.

**Pull request merged**